### PR TITLE
Verify service events [Verifiable events part 3]

### PR DESCRIPTION
**Pull request by m4ksio**
Verification of service events

### CODE REVIEW

File: module/chunks/chunkVerifier_test.go [hunk 1]
```diff
@@ -346,3 +396,34 @@ func (vm *vmMock) Run(ctx fvm.Context, proc fvm.Procedure, led state.View, progr
...
+	if !ok {
+		return fmt.Errorf("invokable is not a transaction")
+	}
+
+	tx.ServiceEvents = []flow.Event{epochSetupEvent}
+
+	_, _ = led.Get("00", "", "")
+	_, _ = led.Get("05", "", "")
+	_ = led.Set("05", "", "", []byte{'B'})
+	tx.Logs = []string{"log1", "log2"}
```
Comment (jordanschalm) [inline module/chunks/chunkVerifier_test.go:413]: What is all this doing? Some comments would help

File: module/chunks/chunkVerifier_test.go [hunk 2]
```diff
@@ -262,15 +301,24 @@ func GetBaselineVerifiableChunk(t *testing.T, script string) *verification.Verif
...
+	erServiceEvents := make([]flow.ServiceEvent, 0)
+
+	if system {
+		chunkEvents = flow.EventsList{}
+		erServiceEvents = serviceEventsList
+	} else {
+		for i := 0; i < collectionSize; i++ {
+			if i == magicTxIndex {
+				switch script {
+				case "failedTx":
```
Comment (jordanschalm) [inline module/chunks/chunkVerifier_test.go:314]: could add a constant for this

File: model/flow/service_event.go [hunk 3]
```diff
@@ -34,6 +24,27 @@ type ServiceEvent struct {
 	Event interface{}
 }
 
+// ServiceEventsList is a handy container to enable comparisons
+type ServiceEventsList []ServiceEvent
```
Comment (jordanschalm) [inline model/flow/service_event.go:28]:
```suggestion
// ServiceEventList is a handy container to enable comparisons
type ServiceEventList []ServiceEvent
```
To match our other `*List` types

File: module/chunks/chunkVerifier_test.go [hunk 4]
```diff
@@ -346,3 +396,34 @@ func (vm *vmMock) Run(ctx fvm.Context, proc fvm.Procedure, led state.View, progr
...
+
+type vmSystemOkMock struct{}
+
+func (vm *vmSystemOkMock) Run(ctx fvm.Context, proc fvm.Procedure, led state.View, programs *programs.Programs) error {
+	tx, ok := proc.(*fvm.TransactionProcedure)
+	if !ok {
+		return fmt.Errorf("invokable is not a transaction")
+	}
+
+	tx.ServiceEvents = []flow.Event{epochSetupEvent}
```
Comment (jordanschalm) [inline module/chunks/chunkVerifier_test.go:408]: It wasn't clear to me that this was why this mock was "OK" and the other one was "Bad". Could rename to be more clear eg:
```go
tx.ServiceEvents = expectedServiceEvents
```

or add a comment

File: module/chunks/chunkVerifier_test.go [hunk 5]
```diff
@@ -346,3 +396,34 @@ func (vm *vmMock) Run(ctx fvm.Context, proc fvm.Procedure, led state.View, progr
...
+
+type vmSystemBadMock struct{}
+
+func (vm *vmSystemBadMock) Run(ctx fvm.Context, proc fvm.Procedure, led state.View, programs *programs.Programs) error {
+	tx, ok := proc.(*fvm.TransactionProcedure)
+	if !ok {
+		return fmt.Errorf("invokable is not a transaction")
+	}
+
+	tx.ServiceEvents = []flow.Event{epochCommitEvent}
```
Comment (jordanschalm) [inline module/chunks/chunkVerifier_test.go:426]:
```suggestion
	// EpochSetup event is expected, but we emit EpochCommit here resulting in a chunk fault
	tx.ServiceEvents = []flow.Event{epochCommitEvent}
```
Same here, a comment noting why this mock is expected to fail

File: module/chunks/chunkVerifier_test.go [hunk 6]
```diff
@@ -136,14 +155,34 @@ func (s *ChunkVerifierTestSuite) TestFailedTx() {
...
+	assert.NotNil(s.T(), vch)
+	_, chFault, err := s.systemBadVerifier.SystemChunkVerify(vch)
+	assert.Nil(s.T(), err)
+	assert.NotNil(s.T(), chFault)
+	assert.IsType(s.T(), &chunksmodels.CFInvalidSystemEventsEmitted{}, chFault)
+}
+
+// TestServiceEventsAreChecked ensures that service events are in fact checked
+func (s *ChunkVerifierTestSuite) TestServiceEventsAreChecked() {
+	vch := GetBaselineVerifiableChunk(s.T(), "serviceEventsMismatch", true)
```
Comment (jordanschalm) [inline module/chunks/chunkVerifier_test.go:179]:
```suggestion
	vch := GetBaselineVerifiableChunk(s.T(), "serviceEventsAreChecked", true)
```

I'm guessing we want the script to match the test here

File: model/flow/epoch_test.go [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+
+	t.Run("empty structures are equal", func(t *testing.T) {
+		a := &flow.ClusterQCVoteData{}
+		b := &flow.ClusterQCVoteData{}
+		require.True(t, a.EqualTo(b))
+		require.True(t, b.EqualTo(a))
+	})
+}
+
+//	t.Run("sig data triggers", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (m4ksio) [reply model/flow/epoch_test.go:27]: Good spot, fixed

File: module/chunks/chunkVerifier_test.go (hunk 1 above)
Comment (m4ksio) [reply module/chunks/chunkVerifier_test.go:413]: Done

File: module/chunks/chunkVerifier_test.go (hunk 2 above)
Comment (m4ksio) [reply module/chunks/chunkVerifier_test.go:314]: Yeah, but whole test relies on those simple strings, I think that's fine to keep it consistent

File: module/chunks/chunkVerifier_test.go (hunk 6 above)
Comment (m4ksio) [reply module/chunks/chunkVerifier_test.go:179]: It doesn't matter really, but better to keep it right

**Review by jordanschalm — APPROVED** [review]

File: model/chunks/chunkFaults.go [hunk 8]
```diff
@@ -108,6 +108,35 @@ func (c *CFInvalidEventsCollection) String() string {
...
+func (c *CFInvalidServiceEventsEmitted) ChunkIndex() uint64 {
+	return c.chunkIndex
+}
+
+func (c *CFInvalidServiceEventsEmitted) ExecutionResultID() flow.Identifier {
+	return c.resultID
+}
+
+func (c *CFInvalidServiceEventsEmitted) String() string {
+	return fmt.Sprintf("service events differs, got [%s] expected [%s] for chunk %d with result ID %s", c.computed, c.expected, c.chunkIndex, c.resultID)
```
Comment (yhassanzadeh13) [inline model/chunks/chunkFaults.go:137]: shouldn't the place holder for result ID be `%x`?

**Review by yhassanzadeh13 — APPROVED** [review]
LGTM!

**Pull request merged**